### DISCUSSION TITLE
How to handle forward and backward pagination properly with GraphQL

### DISCUSSION THREAD

**Discussion by albx** [Azure/data-api-builder#2088] [Q&A]
Hi everyone.
I'm using Data API Builder to create a GraphQL endpoint and I noticed that the schema for a connection has an _endCursor_ property, which indicates the cursor to use for moving forward in the pagination.

However, I'm not sure how to handle backward pagination, because it seems that the schema does not provide any information to move to a previous page.

As far as I know, HotChocolate supports two approaches:
- with cursors, it shows both an _endCursor_ and a _startCursor_ (see https://chillicream.com/docs/hotchocolate/v13/fetching-data/pagination#connections)
- with offset, using _skip_ and _take_ properties (see https://chillicream.com/docs/hotchocolate/v13/fetching-data/pagination/#offset-pagination)

Relay, on the other hand, uses a cursor approach but it allows both forward and backward pagination using an _after_ and a _before_ property, which contain the cursor value (see https://relay.dev/graphql/connections.htm#sec-Forward-pagination-arguments and https://relay.dev/graphql/connections.htm#sec-Backward-pagination-arguments).

Is there is any way to achieve this? Please, let me know if you need more informations.
Thank you very much!

**Comment by frekvent-szabolcs** [comment]
Ran into this as well. If you are using a SQL database, you can define a stored procedure that queries the meta data needed for your pagination.

The `endCursor` is a base64 encoded string. If you decode it, you get something like this:
```
[{"EntityName":"Currency","FieldName":"Code","FieldValue":"RSD","Direction":1}]
```

The direction needs to match your current sorting and the field value specifies after which value you want the page. 

So If you have a sp that takes the PageSize as a parameter and returns something like this (ex PageSize=2):
| PageNum | IsFirstInPage | IsLastInPage | Code |
| ------- | ------------- | ------------ | ---- |
| 1       | 1             | 0            | EUR  |
| 1       | 0             | 1            | GBP  |
| 2       | 1             | 0            | HUF  |
| 2       | 0             | 1            | RON  |
| 3       | 1             | 0            | RSD  |
| 3       | 0             | 1            | USD  |

You can compose the endCursor object yourself and encode it as a base64 string. 

You can call the SP using graphql:
![image](https://github.com/Azure/data-api-builder/assets/135596489/602115c4-a486-4f56-8f89-aa681a1a0e20)

This only works for a fix sorting parameter. You can add a second parameter to the stored procedure and use `ORDER BY CASE WHEN @SortField = 'Column1' ELSE ...` to figure out the pages.

Another thing to consider: If the pagination meta data is too large, you probably need infinite scrolling anyway or If you only want forward and backward buttons, you can store the field values needed inside the client's state (every time you page, you store the previous last page item value).

**Reply by albx** [reply]
Thanks for the reply.
Yes, in the end I stored the cursor in memory, enabling only forward and backward buttons (see https://live.morialberto.it/archivio as a sample (italian language)).

Interesting the Stored procedure, I didn't know the form of the cursor in such a deep way :).

**Reply by Baker-Denn** [reply]
I solved this by creating a Dictionary<int,string?> that stored the cursors as the user paged the results:

```
# Sets defaults values
private int currentPage = 1;
private Dictionary<int, string?> pageCursors = new() { { 1, null } };

# Increments current page and adds cursor to dictionary
private void NextPage(string cursor) => pageCursors[++currentPage] = cursor;

# Decrements current page
private void PreviousPage() => currentPage--;
```
Then I simply pass `pageCursors[currentPage]` as my after value to my query.

**Reply by albx** [reply]
Yep, I did the same thing 😄 (https://github.com/albx/live.morialberto.it/blob/main/src/MoriAlberto.Live.WebSite/Services/PageCursorProvider.cs)

**Comment by JerryNixon** [comment]
We want this too. https://github.com/Azure/data-api-builder/issues/2459

**Comment by JerryNixon** [comment]
I think a programmatic solution to navigate to the previous cursor is going to be the responsibility of the client as the stateless Data API does not know the records in the previous dataset. It would require a considerable amount of increase database traffic to do what you are asking. However, we are wanting to add considerable amounts of pagination data to the response payload to help as much as we can.

**Reply by albx** [reply]
Agree, in my case the solution was storing in memory a dictionary of page index with its cursor. It worked, but I think the pagination metadata could be very useful.
Thank you!
